GoogleCloudPlatform/terraform-validator
# Implemented fetch & merge logic on Converter

This PR implements fetch & mergeDelete functionality in terraform-validator, with one concrete implementation (Project IAM mergeDelete - see https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/631).

Local manual testing:

1. Check out this branch of terraform-validator and the https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/631 branch of terraform-google-conversion
2. Add `replace github.com/GoogleCloudPlatform/terraform-google-conversion v0.0.0-20210105223515-0824b0da180b => /path/to/terraform-google-conversion` to go.mod
3. Run `make build` inside terraform-validator. This will put a binary in `terraform-validator/bin`.
4. Create a terraform file that adds a binding or member to your test project.
5. Apply the file.
6. Remove the binding / member from the file.
7. Run `terraform plan --out tfplan.tfplan`
8. Run `terraform show -json ./tfplan.tfplan > ./tfplan.json`
9. Run `/path/to/terraform-validator/bin/terraform-validator convert ./tfplan.json`

The output of the final command is JSON. You can inspect it to see that your IAM policy (minus the binding / member you removed) is present, despite not being in terraform. You can re-run this process starting at step 3 (but from the terraform-validator master branch) to see the previous behavior (which will give you an empty iam policy.)

~Regarding automated testing: terraform-google-conversion has automated tests of the merge behavior for deletes, but there are no integration tests in this PR for the fetch behavior. I'm not sure how to implement them in the context of terraform-validator, since it isn't set up for VCR tests as far as I know. Please let me know if you have suggestions on how I could implement something!~ See comment below re: automated tests.

Related to https://github.com/hashicorp/terraform-provider-google/issues/7797

This is a breaking change; it looks like the last release was 2021-01-21, right before the changes for supporting IAM better started coming through, so I think we don't need to do a release prior to merging this PR.
